prov/verbs/src/verbs_info.c

```diff
@@ -1207,6 +1209,23 @@ int fi_ibv_init_info(const struct fi_info **all_infos)
 	ep_type[2] = &verbs_dgram_domain;
 
 	if (!fi_ibv_gl_data.fork_unsafe) {
+		/* check to see if hugepages can be used */
+		buffer = mmap(NULL, VERBS_MMAP_TEST_BUFFER_LEN_SIZE,
+			PROT_READ | PROT_WRITE,
+			MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB,
+			-1, 0);
+		if (buffer != MAP_FAILED) {
+			ret = munmap(buffer, VERBS_MMAP_TEST_BUFFER_LEN_SIZE);
+			assert(ret == 0);
+
+			/* assume hugepages will be used and set
+			   environment for correctness */
+                        setenv("RDMAV_HUGEPAGES_SAFE", "1", 1);
+
+                        VERBS_INFO(FI_LOG_CORE, "Huge pages may be in-use, "
+				"enabling huge page support\n");
+                }
```

Comment:
As enabling this by default increases the memory registration latency, I wonder if this should be left to the user to set this env variable if they see a memory registration failure.

Reply:
If they don't enable it, the application with crash with a segmentation fault in the common core code. @swelch was working on a patch for that actually, but at minimum -- we would need some sort of diagnostic/debugging message to the user so that they would be aware that such a variable needed to be set.

Reply:
I see. I thought the memory registration call failed by returning an error.

To me it seems that we need this patch if we're to maintain current behavior. In case we turn off fork safety as the default behavior then we can skip this patch. @shefty thoughts?

If we add this patch, can you please update the log message to say that memory registration latency would be high if fork safety is not disabled? And it should probably be a warning instead of info level log.

Reply:
Sounds fine to me. If we disable the libfabric verbs fork support by default, and provide guidance to application writers to use the libibverbs environment variables -- then I feel that we can discard this patch. That would resolve #4969  and #4974 in one go.

Reply:
If the measurable performance impact is significant or if our fork support doesn't really exist because of other issues, I think we can disable it.  But until we can verify one of those, I think we proceed with this patch to maintain the current behavior as Arun mentioned.

Reply:
In an alltoall between 64 ranks (1 rank per node), I observed a 10x increase in collective latency with 256KB messages and higher. As I mentioned in the original issue, when using the provider as-is today, there is a 30-60% increase in latency for messages above the rendezvous threshold. The increased latency for message sizes above the rendezvous threshold becomes more pronounced as the number of ranks increases. Additionally, if the rendezvous threshold is decreased, all messages above the rendezvous threshold experience additionally latency. 

When running with hugepages, and RDMAV_HUGEPAGES_SAFE=1 for hugepage safety, and default fork safety today -- there is a performance impact to applications between .3 to 10x or more. 

It doesn't affect the provider as much if the memory registration cache is in use, but we do not use it today due to silent data corruption issues. #4946 is relevant to that discussion.

Reply:
Is the performance impact related to huge pages or fork?

Reply:
I think this is the answer I'm looking for
```
default fork safety today - there is a performance impact to applications between .3 to 10x or more.
```
That delta is based on disabling for support, correct?

Reply:
To be completely accurate, 

```
If 
    the provider requests fork safety, (FI_FORK_UNSAFE=0, [default]) and 
    hugepages safety IS NOT requested (RDMAV_HUGEPAGES_SAFE=0), 
        latency is normal. [ NOTE: this condition requires that hugepages are not enabled, or registration errors occur ]

If 
    the provider requests fork safety, (FI_FORK_UNSAFE=0, [default]) and 
    hugepages safety IS requested (RDMAV_HUGEPAGES_SAFE=1), 
        latency increases by a factor of .3 to 10x or more

If 
    the provider does not request fork safety (FI_FORK_UNSAFE=1) , 
    latency is normal. [RMDAV_HUGEPAGES_SAFE is ignored in this case] 
```

The delta in performance is really associated with enabling hugepage safety if libibverbs fork support is requested. If libibverbs fork support isn't requested, hugepage checking isn't necessary in libibverbs.

Reply:
This helps.  So the change from our original decision is that hugepages are more of a factor now and can have a significant impact.  Thanks!

@a-ilango @jsquyres @hppritcha -The short question here is whether we should disable fork support by default now, to greatly improve performance when huge pages are in use.

Reply:
I would be in favor of removing the FI_FORK_SAFE variable and disabling fork support by default. The other issue, #4974, indicates a problem in multi-model applications where different libraries calling into libibverbs can complicate matters, or cause errors. 

As much as I dislike the idea of applications using the IBV/RDMAV environment variables directly, it doesn't seem like we can prevent a bad order of operations to the libibverbs layer without extracting ourselves from environment variable management. The IBV/RDMAV environment variables are "global" in the sense that it doesn't matter who makes it to libibverbs first. FI_* variables will influence the underlying libraries at the point in time where the provider is initialized, which could be after the application probes ibverbs by itself. 

Applications using the verbs provider, or intending to use the verbs provider can set IBV_FORK_SAFE, and/or RDMAV_HUGEPAGES_SAFE with the correct context. Libfabric has no idea whether an application will attempt to call ibv_* functions directly, but the application does.

Reply:
@jswaro  I agree.  I don't see a way for libfabric to use ibv_fork_init() safely.  I think we need to remove the fork support from libfabric and rely on it being set directly when verbs is used.

Reply:
This kinda sucks, but I don't see a better answer, either.

This needs to be clearly documented somewhere.

Reply:
@jsquyres Agreed. I was thinking we should potentially add another sections 'notes' or similar, for the purpose of discussing additional environment variable that influence the provider, but fall outside of the domain of libfabric such as IBV_FORK_SAFE, or RDMAV_HUGEPAGES_SAFE. It could also be a section for discussing information relevant to a provider, but not directly influenced by the provider. 

That same section could be used for MR cache change, #4946 & #4956, to describe the limitations of the implementation with respect to the OS. In that specific case, something like this could go into a 'notes' section:

> The support for hugetlbfs and shared memory areas and non-page-fault
>        events was added in Linux 4.11